Image pre-processing

Foxman13 opened the pull request.

Commit:
built a pre-processor that equalizes an image's contrast histogram

Commit:
added .gitignore for test_data

Commit:
remove test_data

c-w commented on preprocessor/main.py:
```diff
@@ -0,0 +1,30 @@
+    outputdirectory = ''
+
+    try:
+        opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
```
NB: the Python docs recommend to use [argparse](https://docs.python.org/3/library/argparse.html) over getopt and optparse:

> Note that an equivalent command line interface could be produced with less code and more informative help and error messages by using the argparse module

From [getopt docs](https://docs.python.org/3/library/getopt.html).

c-w commented on preprocessor/main.py:
```diff
@@ -0,0 +1,30 @@
+    try:
+        opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
+    except getopt.GetoptError:
+        print('test.py -i <inputdirectory> -o <outputdirectory>')
```
Should this be `main.py` since that's the script name now? Alternatively, `sys.argv[0]` has the name of the script dynamically.

c-w commented on preprocessor/main.py:
```diff
@@ -0,0 +1,30 @@
+    normalize_images(inputdirectory, outputdirectory)
+
+if __name__== "__main__":
+    main(sys.argv[1:])
```
Nit: missing newline at end of file. Here are some other lint errors:

*flake8*

```
preprocessor/main.py:1:1: F401 'os.path.join' imported but unused
preprocessor/main.py:1:1: F401 'os.path.exists' imported but unused
preprocessor/main.py:2:11: E401 multiple imports on one line
preprocessor/main.py:5:1: E302 expected 2 blank lines, found 1
preprocessor/main.py:10:40: E231 missing whitespace after ','
preprocessor/main.py:10:48: E231 missing whitespace after ','
preprocessor/main.py:10:58: E231 missing whitespace after ','
preprocessor/main.py:16:7: E111 indentation is not a multiple of four
preprocessor/main.py:17:10: E111 indentation is not a multiple of four
preprocessor/main.py:18:10: E111 indentation is not a multiple of four
preprocessor/main.py:19:7: E111 indentation is not a multiple of four
preprocessor/main.py:20:10: E111 indentation is not a multiple of four
preprocessor/main.py:21:7: E111 indentation is not a multiple of four
preprocessor/main.py:22:10: E111 indentation is not a multiple of four
preprocessor/main.py:29:1: E305 expected 2 blank lines after class or function definition, found 1
preprocessor/main.py:29:12: E225 missing whitespace around operator
preprocessor/main.py:30:23: W292 no newline at end of file
preprocessor/processor.py:2:1: F401 'os' imported but unused
preprocessor/processor.py:3:1: F401 'glob' imported but unused
preprocessor/processor.py:8:80: E501 line too long (80 > 79 characters)
preprocessor/processor.py:9:1: E302 expected 2 blank lines, found 1
preprocessor/processor.py:29:1: E302 expected 2 blank lines, found 1
preprocessor/processor.py:33:18: E231 missing whitespace after ','
preprocessor/processor.py:33:20: E231 missing whitespace after ','
preprocessor/processor.py:33:51: E231 missing whitespace after ','
preprocessor/processor.py:33:53: E231 missing whitespace after ','
preprocessor/processor.py:38:1: E302 expected 2 blank lines, found 1
preprocessor/processor.py:41:1: E302 expected 2 blank lines, found 1
preprocessor/processor.py:52:1: E302 expected 2 blank lines, found 1
preprocessor/processor.py:56:42: W292 no newline at end of file
```

*pylint*

```
************* Module main
preprocessor/main.py:10:39: C0326: Exactly one space required after comma
        opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
                                       ^ (bad-whitespace)
preprocessor/main.py:10:47: C0326: Exactly one space required after comma
        opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
                                               ^ (bad-whitespace)
preprocessor/main.py:10:57: C0326: Exactly one space required after comma
        opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
                                                         ^ (bad-whitespace)
preprocessor/main.py:16:0: W0311: Bad indentation. Found 6 spaces, expected 8 (bad-indentation)
preprocessor/main.py:17:0: W0311: Bad indentation. Found 9 spaces, expected 12 (bad-indentation)
preprocessor/main.py:18:0: W0311: Bad indentation. Found 9 spaces, expected 12 (bad-indentation)
preprocessor/main.py:19:0: W0311: Bad indentation. Found 6 spaces, expected 8 (bad-indentation)
preprocessor/main.py:20:0: W0311: Bad indentation. Found 9 spaces, expected 12 (bad-indentation)
preprocessor/main.py:21:0: W0311: Bad indentation. Found 6 spaces, expected 8 (bad-indentation)
preprocessor/main.py:22:0: W0311: Bad indentation. Found 9 spaces, expected 12 (bad-indentation)
preprocessor/main.py:29:11: C0326: Exactly one space required before comparison
if __name__== "__main__":
           ^^ (bad-whitespace)
preprocessor/main.py:30:0: C0304: Final newline missing (missing-final-newline)
preprocessor/main.py:2:0: C0410: Multiple imports on one line (sys, getopt) (multiple-imports)
preprocessor/main.py:10:14: W0612: Unused variable 'args' (unused-variable)
preprocessor/main.py:1:0: W0611: Unused join imported from os.path (unused-import)
preprocessor/main.py:1:0: W0611: Unused exists imported from os.path (unused-import)
************* Module processor
preprocessor/processor.py:30:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:31:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:33:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:33:17: C0326: Exactly one space required after comma
        img_yuv[:,:,0] = cv.equalizeHist(img_yuv[:,:,0])
                 ^ (bad-whitespace)
preprocessor/processor.py:33:19: C0326: Exactly one space required after comma
        img_yuv[:,:,0] = cv.equalizeHist(img_yuv[:,:,0])
                   ^ (bad-whitespace)
preprocessor/processor.py:33:50: C0326: Exactly one space required after comma
        img_yuv[:,:,0] = cv.equalizeHist(img_yuv[:,:,0])
                                                  ^ (bad-whitespace)
preprocessor/processor.py:33:52: C0326: Exactly one space required after comma
        img_yuv[:,:,0] = cv.equalizeHist(img_yuv[:,:,0])
                                                    ^ (bad-whitespace)
preprocessor/processor.py:35:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:36:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:42:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:44:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:45:0: W0311: Bad indentation. Found 12 spaces, expected 8 (bad-indentation)
preprocessor/processor.py:46:0: W0311: Bad indentation. Found 12 spaces, expected 8 (bad-indentation)
preprocessor/processor.py:47:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:48:0: W0311: Bad indentation. Found 12 spaces, expected 8 (bad-indentation)
preprocessor/processor.py:50:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:53:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:55:0: W0311: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
preprocessor/processor.py:56:0: C0304: Final newline missing (missing-final-newline)
preprocessor/processor.py:56:0: W0311: Bad indentation. Found 12 spaces, expected 8 (bad-indentation)
preprocessor/processor.py:4:0: E0401: Unable to import 'cv2' (import-error)
preprocessor/processor.py:2:0: W0611: Unused import os (unused-import)
preprocessor/processor.py:3:0: W0611: Unused import glob (unused-import)
```

I'll take an item to enable CI->lint for the supporting tools (so not just the main app) after I'm done with the FaceAPI integration.

c-w commented on preprocessor/.gitignore:
```diff
@@ -0,0 +1 @@
+test_data
```
Please move the script to `scripts/preprocessor` so that we have all the scripts in the same place. Note that the validation script should also be moved there since it's currently in the wrong place and @michaelperel has an action item for that from a past pull request review.

c-w commented on preprocessor/processor.py:
```diff
@@ -0,0 +1,56 @@
+from pathlib import Path
+import os
+import glob
+import cv2 as cv
```
Please also add a `requirements.txt` that lists the version of cv2 that the script needs.

c-w commented on preprocessor/processor.py:
```diff
@@ -0,0 +1,56 @@
+        abs_file_path = Path.absolute(f)
+        file_path = str(abs_file_path)
+        output_file_name = get_output_file_name(f, inputpath, outputpath)
+        print('normalizing image: ' + str(f) + ' to: ' + str(output_file_name))
```
Nit: for consistency with `main.py`, it would be good to use string formatting here:

```py
print('normalizing image: {} to: {}'.format(f, output_file_name))
```

c-w commented on preprocessor/processor.py:
```diff
@@ -0,0 +1,56 @@
+        img = cv.imread(file_path)
+        img_yuv = cv.cvtColor(img, cv.COLOR_BGR2YUV)
+
+        img_yuv[:,:,0] = cv.equalizeHist(img_yuv[:,:,0])
```
I wonder if there are some other cleanups that we should also implement here. I haven't done much research in this area yet, but [skimage.expose](http://scikit-image.org/docs/dev/api/skimage.exposure.html) has a bunch of nice utilities. If you have time, it would be great to investigate what other sorts of cleanups we could/should perform.

Commit:
implemented PR review changes

-moved to scripts directory
-changed from opencv to scikit-image adaptive histogam equalization
-changed getopt to argparse

Commit:
update to docker file

use requirements.txt
